Convert the `Snackbar` component to TypeScript

kienstra opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Convert the `Snackbar` component to TypeScript

## Why?
As part of an [effort](https://github.com/WordPress/gutenberg/issues/35744) to convert `@wordpress/components` to TypeScript

## How?
Mainly by adding types to the `Snackbar` and `SnackbarList` components

## Testing Instructions
1. `npm run storybook:dev`
2. Expected: The Storybook still works

## Screenshots or screencast <!-- if applicable -->
<img width="1015" alt="Screen Shot 2022-11-01 at 3 56 43 PM" src="https://user-images.githubusercontent.com/4063887/199349991-afd841f5-4873-4b2b-a4e5-5fbf1d0caaf4.png">

## Follow-up

 - `SnackbarList` is not very well-documented (it doesn't even have a README) and prior to this PR, didn't have Storybook examples either. We should re-organize the folder structure, improve the docs and come up with a better Storybook strategy (e.g. use subcomponents and merge `Snackbar` and `SnackbarList` in the same file)
 - We should  refactor the `Notice` component to TypeScript and re-use its types in `Snackbar`
 - I came across a framer-motion bug when clicking rapidly on two `Snackbar`s in a list — the second `Snackbar` goes transparent but doesn't get fully removed
 - The timeout logic for `Snackbar` to be dismissed works in a bit of an unexpected way (snackbar items don't get dismissed in order
 - Take `isDismissible` prop from notice data is not taken into account by `Snackbar` (although this could be on purpose)
 - The `Snackbar` icon (passed through its icon prop) is not vertically centered, doesn't seem to work well with WordPress icons

kienstra committed:
Refactor Snackbar component to TypeScript

kienstra committed:
Merge branch 'trunk' into update/snackbar-to-ts

kienstra committed:
Add a CHANGELOG entry for this PR

kienstra commented on packages/components/src/snackbar/list.tsx:
```diff
@@ -40,27 +43,33 @@ const SNACKBAR_VARIANTS = {
-	init: false,
-	open: false,
-	exit: false,
+	init: {},
```
`false` caused a TypeScript error: 
<img width="852" alt="Screen Shot 2022-11-01 at 4 17 47 PM" src="https://user-images.githubusercontent.com/4063887/199352682-467c6aee-1676-4c5c-ba76-213c58d1bd6f.png">

kienstra commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,116 @@
+	onClick?: ( event: ButtonEvent ) => void;
+};
+
+export type Notice = {
```
It'd be better to import [WPNotice](https://github.com/WordPress/gutenberg/blob/62987d7e995a6d24f6b7d5a491c73aff5a1097c8/packages/notices/src/store/selectors.js#L20) from `packages/notices/`, instead of defining this `Notice` type.

But I couldn't find a way to import from `packages/notices/`, given that it's JSDoc, and it doesn't export a type.

ciampo replied:
The `@wordpress/notices` package is not currently build with the TypeScript compiler, and therefore it's not currently possible to import the `WPNotice` type.

But even if it was possible, I would still be unsure about adding a depending in the components package from a "Gutenberg App"-related package — we'd like to keep the components package as decoupled as possible from the Gutenberg app.

Let's keep the `Notice` type in this file for now.

On top of that, let's add a comment about these types potentially being moved to the `Notice` component once it gets typed

`// TODO: move this type to the Notice component once it gets typed`

kienstra replied:
Good idea, I committed the TODO in https://github.com/WordPress/gutenberg/pull/45472/commits/5636934be887f7efdc19536ec4e6f5c749ab15fa

ciampo replied:
Apologies for not being 100% clear — I was proposing to add the TODO comment in `packages/components/src/snackbar/types.ts`, above the `NoticeAction` type. We don't want to touch the types in `@wordpress/notices`, because those types are for the data store, and should therefore not be related to the component.


Would you mind moving the comment over? Thank you!

kienstra replied:
Ah, sorry. Moved it in https://github.com/WordPress/gutenberg/pull/45472/commits/9af48f2ed27ce98a5e531f05909a85dcc7aab75b

kienstra committed:
Reference SnackbarProps for message and politeness

kienstra commented on packages/components/src/snackbar/README.md:
```diff
@@ -36,17 +36,54 @@ const MySnackbarNotice = () => (
 The following props are used to control the display of the component.
 
--   `children`: (string) The displayed message of a notice. Also used as the spoken message for assistive technology, unless `spokenMessage` is provided as an alternative message.
--   `spokenMessage`: (string) Used to provide a custom spoken message in place of the `children` default.
```
I changed the type of `spokenMessage` from `string` to `ReactNode`.

The type `ReactNode` includes `string`.

And the fact that [children is the default of spokenMessage](https://github.com/kienstra/gutenberg/blob/62987d7e995a6d24f6b7d5a491c73aff5a1097c8/packages/components/src/snackbar/index.js#L46) was a problem:

```jsx
function Snackbar(
	{
		className,
		children,
		spokenMessage = children,
```

The type of `children`, `ReactNode`, isn't assignable to `string`:

<img width="861" alt="Screen Shot 2022-11-01 at 4 34 06 PM" src="https://user-images.githubusercontent.com/4063887/199356175-350fc371-090a-40f7-9c76-dae70b3d2a8d.png">

ciampo replied:
Looking at the old component's docs, `children` used to be of type `string` too. We should therefore mark both `children` and `spokenMessage` as `string`, especially because `spokenMessage` comes from the `Notice` type, which should be representing some kind of raw data, and should not be related to React.

kienstra replied:
Thanks, https://github.com/WordPress/gutenberg/pull/45472/commits/7b0c570fac9e8c67b41c3b32752e5f710d77188b changes both to `string`.

kienstra committed:
Wrap the politness type with NonNullable

kienstra commented:
Hi @ciampo,
Would you have time to review this? Thanks so much for helping me with the `ColorPalette` one.

Thanks, Marco!

ciampo commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,116 @@
+	id: string;
+	spokenMessage: ReactNode;
+	actions: Action[];
+	icon?: WPElement | null;
```
I wonder if we could use `ReactNode` here instead of `WPElement`

kienstra replied:
Sure, that's cleaner. Changed in https://github.com/WordPress/gutenberg/pull/45472/commits/ecb238793f1f05b3677350ff89ce912e7c232fd6.

ciampo commented on packages/components/src/snackbar/list.tsx:
```diff
@@ -40,27 +43,33 @@ const SNACKBAR_VARIANTS = {
+	children,
+	onRemove = noop,
+}: WordPressComponentProps< SnackbarListProps, 'div' > ) {
+	const listRef = useRef() as MutableRefObject< HTMLDivElement >;
```
Let's not override the return type of `useRef` with the `as` keyword — instead, we can specific the type of ref with the `useRef< RefType >( ... )` syntax.

Doing so highlights two more issues that the `as` override was hiding:
- in this file, when passing the `ref` to the `<div>`, there's a typescript error caused by the fact that the ref's value can also be `null`. We can solve that easily by changing the starting value of the `ref` from `undefined` to `null`, and by adding `null` to the ref's type
- in `index.tsx`, when calling `listRef?.current.focus()`, there is the chance that `listRef.current` is still `undefined`

Finally, we should be a bit less specific when declaring the type of `listRef` in `types.ts` — using `HTMLElement` instead of `HTMLDivElement` should be enough, while allowing us to change the element representing the list from a `div` to any other HTML element.

```diff
diff --git a/packages/components/src/snackbar/index.tsx b/packages/components/src/snackbar/index.tsx
index 84ee319ad5..f2b223276d 100644
--- a/packages/components/src/snackbar/index.tsx
+++ b/packages/components/src/snackbar/index.tsx
@@ -70,7 +70,7 @@ function UnforwardedSnackbar(
 		}
 
 		// Prevent focus loss by moving it to the list element.
-		listRef?.current.focus();
+		listRef?.current?.focus();
 
 		onDismiss();
 		onRemove();
diff --git a/packages/components/src/snackbar/list.tsx b/packages/components/src/snackbar/list.tsx
index a98bbc5bc1..b073a251bb 100644
--- a/packages/components/src/snackbar/list.tsx
+++ b/packages/components/src/snackbar/list.tsx
@@ -1,7 +1,6 @@
 /**
  * External dependencies
  */
-import type { MutableRefObject } from 'react';
 import classnames from 'classnames';
 
 /**
@@ -66,7 +65,7 @@ export function SnackbarList( {
 	children,
 	onRemove = noop,
 }: WordPressComponentProps< SnackbarListProps, 'div' > ) {
-	const listRef = useRef() as MutableRefObject< HTMLDivElement >;
+	const listRef = useRef< HTMLDivElement | null >( null );
 	const isReducedMotion = useReducedMotion();
 	className = classnames( 'components-snackbar-list', className );
 	const removeNotice = ( notice: Notice ) => () => onRemove( notice.id );
diff --git a/packages/components/src/snackbar/types.ts b/packages/components/src/snackbar/types.ts
index 502b6b27ab..52d8eb534b 100644
--- a/packages/components/src/snackbar/types.ts
+++ b/packages/components/src/snackbar/types.ts
@@ -95,7 +95,7 @@ export type SnackbarProps = {
 	/**
 	 * A ref to the list that contains the snackbar.
 	 */
-	listRef?: MutableRefObject< HTMLDivElement >;
+	listRef?: MutableRefObject< HTMLElement | null >;
 };
 
 export type SnackbarListProps = {
```

kienstra replied:
Good idea, better to set the correct type than cast it.

Committed in https://github.com/WordPress/gutenberg/pull/45472/commits/cef1524fca812d5d51a8706f4b40155327695109

ciampo commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,116 @@
+	/**
+	 * Function called when a notice should be removed / dismissed.
+	 *
+	 * @default () => void
```
Actually, no need to keep the `noop` default — let's clean up the code a little bit, and use optional chaining syntax instead

<details>

<summary>Like so</summary>

```diff
diff --git a/packages/components/src/snackbar/index.tsx b/packages/components/src/snackbar/index.tsx
index 84ee319ad5..70eb1ed2e5 100644
--- a/packages/components/src/snackbar/index.tsx
+++ b/packages/components/src/snackbar/index.tsx
@@ -19,7 +19,6 @@ import Button from '../button';
 import type { Action, ButtonEvent, SnackbarProps } from './types';
 import type { WordPressComponentProps } from '../ui/context';
 
-const noop = () => {};
 const NOTICE_TIMEOUT = 10000;
 
 /**
@@ -47,19 +46,17 @@ function UnforwardedSnackbar(
 		spokenMessage = children,
 		politeness = 'polite',
 		actions = [],
-		onRemove = noop,
+		onRemove,
 		icon = null,
 		explicitDismiss = false,
 		// onDismiss is a callback executed when the snackbar is dismissed.
 		// It is distinct from onRemove, which _looks_ like a callback but is
 		// actually the function to call to remove the snackbar from the UI.
-		onDismiss = noop,
+		onDismiss,
 		listRef,
 	}: WordPressComponentProps< SnackbarProps, 'div' >,
 	ref: ForwardedRef< any >
 ) {
-	onDismiss = onDismiss || noop;
-
 	function dismissMe(
 		event:
 			| KeyboardEvent< HTMLDivElement | HTMLSpanElement >
@@ -72,14 +69,14 @@ function UnforwardedSnackbar(
 		// Prevent focus loss by moving it to the list element.
 		listRef?.current.focus();
 
-		onDismiss();
-		onRemove();
+		onDismiss?.();
+		onRemove?.();
 	}
 
 	function onActionClick( event: ButtonEvent, onClick: Action[ 'onClick' ] ) {
 		event.stopPropagation();
 
-		onRemove();
+		onRemove?.();
 
 		if ( onClick ) {
 			onClick( event );
@@ -92,8 +89,8 @@ function UnforwardedSnackbar(
 	useEffect( () => {
 		const timeoutHandle = setTimeout( () => {
 			if ( ! explicitDismiss ) {
-				onDismiss();
-				onRemove();
+				onDismiss?.();
+				onRemove?.();
 			}
 		}, NOTICE_TIMEOUT );
 
@@ -123,10 +120,10 @@ function UnforwardedSnackbar(
 		<div
 			ref={ ref }
 			className={ classes }
-			onClick={ ! explicitDismiss ? dismissMe : noop }
+			onClick={ ! explicitDismiss ? dismissMe : undefined }
 			tabIndex={ 0 }
 			role={ ! explicitDismiss ? 'button' : '' }
-			onKeyPress={ ! explicitDismiss ? dismissMe : noop }
+			onKeyPress={ ! explicitDismiss ? dismissMe : undefined }
 			aria-label={ ! explicitDismiss ? __( 'Dismiss this notice' ) : '' }
 		>
 			<div className={ snackbarContentClassnames }>
diff --git a/packages/components/src/snackbar/list.tsx b/packages/components/src/snackbar/list.tsx
index a98bbc5bc1..1b49016a95 100644
--- a/packages/components/src/snackbar/list.tsx
+++ b/packages/components/src/snackbar/list.tsx
@@ -21,7 +21,6 @@ import {
 import type { Notice, SnackbarListProps } from './types';
 import type { WordPressComponentProps } from '../ui/context';
 
-const noop = () => {};
 const SNACKBAR_VARIANTS = {
 	init: {
 		height: 0,
@@ -64,12 +63,12 @@ export function SnackbarList( {
 	notices,
 	className,
 	children,
-	onRemove = noop,
+	onRemove,
 }: WordPressComponentProps< SnackbarListProps, 'div' > ) {
 	const listRef = useRef() as MutableRefObject< HTMLDivElement >;
 	const isReducedMotion = useReducedMotion();
 	className = classnames( 'components-snackbar-list', className );
-	const removeNotice = ( notice: Notice ) => () => onRemove( notice.id );
+	const removeNotice = ( notice: Notice ) => () => onRemove?.( notice.id );
 	return (
 		<div className={ className } tabIndex={ -1 } ref={ listRef }>
 			{ children }
diff --git a/packages/components/src/snackbar/types.ts b/packages/components/src/snackbar/types.ts
index 502b6b27ab..b21b915e87 100644
--- a/packages/components/src/snackbar/types.ts
+++ b/packages/components/src/snackbar/types.ts
@@ -65,8 +65,6 @@ export type SnackbarProps = {
 	actions?: Notice[ 'actions' ];
 	/**
 	 * Called to remove the snackbar from the UI.
-	 *
-	 * @default () => void
 	 */
 	onRemove?: Function;
 	/**
@@ -88,8 +86,6 @@ export type SnackbarProps = {
 	 *
 	 * It is distinct from onRemove, which _looks_ like a callback but is
 	 * actually the function to call to remove the snackbar from the UI.
-	 *
-	 * @default () => void
 	 */
 	onDismiss?: Notice[ 'onDismiss' ];
 	/**
@@ -109,8 +105,6 @@ export type SnackbarListProps = {
 	children?: ReactNode;
 	/**
 	 * Function called when a notice should be removed / dismissed.
-	 *
-	 * @default () => void
 	 */
 	onRemove?: Function;
 };
```
</details>

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/45472/commits/202787848cb96ce57a4c683a342de016e29add8c

ciampo commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,116 @@
+
+export type Action = {
+	label: string;
+	url: string;
```
It looks like [`url` is optional](https://github.com/WordPress/gutenberg/blob/62987d7e995a6d24f6b7d5a491c73aff5a1097c8/packages/notices/src/store/actions.js#L10) in the original `WPNoticeAction` type.

But also, the old README said "Each member object should contain a `label` and either a `url` link string or `onClick` callback function" — from the sound of this, it looks like a more correct type could be

```diff
diff --git a/packages/components/src/snackbar/types.ts b/packages/components/src/snackbar/types.ts
index 502b6b27ab..cadc8aa7b4 100644
--- a/packages/components/src/snackbar/types.ts
+++ b/packages/components/src/snackbar/types.ts
@@ -17,12 +17,20 @@ export type ButtonEvent =
 	| KeyboardEvent< HTMLButtonElement >
 	| MouseEvent< HTMLButtonElement >;
 
-export type Action = {
+type NoticeActionWithURL = {
 	label: string;
 	url: string;
 	onClick?: ( event: ButtonEvent ) => void;
 };
 
+type NoticeActionWithOnClick = {
+	label: string;
+	url?: string;
+	onClick: ( event: ButtonEvent ) => void;
+};
+
+export type Action = NoticeActionWithURL | NoticeActionWithOnClick;
+
 export type Notice = {
 	id: string;
 	spokenMessage: ReactNode;
```

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/45472/commits/50f5f9827e99816a837c6bf53fb3ee87a167ee35

ciampo commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,116 @@
+	| KeyboardEvent< HTMLButtonElement >
+	| MouseEvent< HTMLButtonElement >;
+
+export type Action = {
```
`NoticeAction` seems like a better name for this type

kienstra replied:
Good idea, renamed in https://github.com/WordPress/gutenberg/pull/45472/commits/72fdce954661efa7b21dc3076db81bae0fe4916c

ciampo commented on packages/components/src/snackbar/index.tsx:
```diff
@@ -54,24 +55,28 @@ function Snackbar(
+		event:
+			| KeyboardEvent< HTMLDivElement | HTMLSpanElement >
+			| MouseEvent< HTMLDivElement | HTMLSpanElement >
+	) {
```
No need to be this specific with the HTML element being clicked, a generic mouse or keybaord event will do

```suggestion
	function dismissMe( event: KeyboardEvent | MouseEvent ) {
```

ciampo commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,116 @@
+	spokenMessage: ReactNode;
+	actions: Action[];
+	icon?: WPElement | null;
+	onDismiss?: Function;
```
Let's avoid using the `Function` type — we can use `() => void` instead

kienstra replied:
Sure, committed in https://github.com/WordPress/gutenberg/pull/45472/commits/20056c37459277ea495c48199b2728fe6372083e

ciampo commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,116 @@
+
+export type ButtonEvent =
+	| KeyboardEvent< HTMLButtonElement >
+	| MouseEvent< HTMLButtonElement >;
```
This type is being used for the `onClick` event on the `Button` component — the `Button` component, when the `href` attribute is passed, renders an `HTMLAnchorElement` and not a `HTMLButtonElement` (this error is not flagged by TypeScript because the `Button` component is not typed yet).

We can actually avoid using this `ButtonEvent` type altogether if we slightly relax the event-related types, using the more generic `SyntheticEvent`

<details>

<summary>Click to expand</summary>

```diff
diff --git a/packages/components/src/snackbar/index.tsx b/packages/components/src/snackbar/index.tsx
index 84ee319ad5..268762a98e 100644
--- a/packages/components/src/snackbar/index.tsx
+++ b/packages/components/src/snackbar/index.tsx
@@ -16,7 +16,7 @@ import warning from '@wordpress/warning';
  * Internal dependencies
  */
 import Button from '../button';
-import type { Action, ButtonEvent, SnackbarProps } from './types';
+import type { Action, SnackbarProps } from './types';
 import type { WordPressComponentProps } from '../ui/context';
 
 const noop = () => {};
@@ -76,7 +76,7 @@ function UnforwardedSnackbar(
 		onRemove();
 	}
 
-	function onActionClick( event: ButtonEvent, onClick: Action[ 'onClick' ] ) {
+	function onActionClick( event: MouseEvent, onClick: Action[ 'onClick' ] ) {
 		event.stopPropagation();
 
 		onRemove();
@@ -140,7 +140,7 @@ function UnforwardedSnackbar(
 							key={ index }
 							href={ url }
 							variant="tertiary"
-							onClick={ ( event: ButtonEvent ) =>
+							onClick={ ( event: MouseEvent ) =>
 								onActionClick( event, onClick )
 							}
 							className="components-snackbar__action"
diff --git a/packages/components/src/snackbar/types.ts b/packages/components/src/snackbar/types.ts
index 502b6b27ab..4856585c6d 100644
--- a/packages/components/src/snackbar/types.ts
+++ b/packages/components/src/snackbar/types.ts
@@ -1,26 +1,17 @@
 /**
  * External dependencies
  */
-import type {
-	KeyboardEvent,
-	MouseEvent,
-	MutableRefObject,
-	ReactNode,
-} from 'react';
+import type { MutableRefObject, ReactNode, SyntheticEvent } from 'react';
 
 /**
  * WordPress dependencies
  */
 import type { WPElement } from '@wordpress/element';
 
-export type ButtonEvent =
-	| KeyboardEvent< HTMLButtonElement >
-	| MouseEvent< HTMLButtonElement >;
-
 export type Action = {
 	label: string;
 	url: string;
-	onClick?: ( event: ButtonEvent ) => void;
+	onClick?: ( event: SyntheticEvent ) => void;
 };
 
 export type Notice = {
```

</details>

kienstra replied:
Nice, that's cleaner without `ButtonEvent`.

Committed in https://github.com/WordPress/gutenberg/pull/45472/commits/e3d2448b3cf9e098dc2b2648f7e097e4a9467b2e

ciampo commented on packages/components/src/snackbar/list.tsx:
```diff
@@ -40,27 +43,33 @@ const SNACKBAR_VARIANTS = {
-	init: false,
-	open: false,
-	exit: false,
+	init: {},
+	open: {},
+	exit: {},
```
Have you tried testing that these new values work as expected for `reducedMotion` users (i.e. the notices don't animate in/out) ?

A cleaner alternative to passing `{}` could be to remove the `SNACKBAR_REDUCE_MOTION_VARIANTS` variable altogether and simply pass `undefined` to the `variant` prop of `motion.div` when `isReducedMotion === true`

kienstra replied:
>Have you tried testing that these new values work as expected for reducedMotion users (i.e. the notices don't animate in/out) ?

I just tested it, and it looks like they don't animate in/out:
![snackbarlist-no-animate](https://user-images.githubusercontent.com/4063887/199592516-f79dde99-8b0f-475b-9be4-6bd703738200.gif)

If it's alright, I just [added](https://github.com/kienstra/gutenberg/blob/9818ca5ff6b337fffb8eb2283916a572be12fe69/packages/components/src/snackbar/stories/list.tsx) a story for `SnackbarList`.

kienstra replied:
> A cleaner alternative to passing {} could be to remove the SNACKBAR_REDUCE_MOTION_VARIANTS variable altogether and simply pass undefined to the variant prop of motion.div when isReducedMotion === true

Thanks, that's much cleaner than keeping an object with 3 properties of `{}`.

Removed `SNACKBAR_REDUCE_MOTION_VARIANTS` in https://github.com/WordPress/gutenberg/pull/45472/commits/8667eb15b12f57617a935b49bda42ad44917f5fb

ciampo replied:
> If it's alright, I just [added](https://github.com/kienstra/gutenberg/blob/9818ca5ff6b337fffb8eb2283916a572be12fe69/packages/components/src/snackbar/stories/list.tsx) a story for `SnackbarList`

Thank you! 

<details>

<summary>I went ahead and made a few additions locally, in order to be able to test the manual dismissal of the notices and their animation (you may want to commit these changes)</summary>

```diff
diff --git a/packages/components/src/snackbar/stories/list.tsx b/packages/components/src/snackbar/stories/list.tsx
index d6567f3feb..a12b2c171a 100644
--- a/packages/components/src/snackbar/stories/list.tsx
+++ b/packages/components/src/snackbar/stories/list.tsx
@@ -3,6 +3,11 @@
  */
 import type { ComponentMeta, ComponentStory } from '@storybook/react';
 
+/**
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
+
 /**
  * Internal dependencies
  */
@@ -29,9 +34,26 @@ export default meta;
 
 export const Default: ComponentStory< typeof SnackbarList > = ( {
 	children,
+	notices: noticesProp,
 	...props
 } ) => {
-	return <SnackbarList { ...props }>{ children }</SnackbarList>;
+	const [ notices, setNotices ] = useState( noticesProp );
+
+	const onRemove = ( id: string ) => {
+		const matchIndex = notices.findIndex( ( n ) => n.id === id );
+		if ( matchIndex > -1 ) {
+			setNotices( [
+				...notices.slice( 0, matchIndex ),
+				...notices.slice( matchIndex + 1 ),
+			] );
+		}
+	};
+
+	return (
+		<SnackbarList { ...props } notices={ notices } onRemove={ onRemove }>
+			{ children }
+		</SnackbarList>
+	);
 };
 
 Default.args = {
@@ -39,7 +61,7 @@ Default.args = {
 		'Use SnackbarList to communicate multiple low priority, non-interruptive messages to the user.',
 	notices: [
 		{
-			id: 'SAVE_POST_NOTICE_ID',
+			id: 'SAVE_POST_NOTICE_ID_1',
 			spokenMessage: 'Post published.',
 			actions: [
 				{
@@ -52,7 +74,7 @@ Default.args = {
 			explicitDismiss: false,
 		},
 		{
-			id: 'SAVE_POST_NOTICE_ID',
+			id: 'SAVE_POST_NOTICE_ID_2',
 			spokenMessage: 'Post updated',
 			actions: [
 				{
```
</details>

Let's review the Storybook examples later (maybe even in a different PRs) — but this is already a small improvement!

kienstra replied:
Oh thanks! Committed this verbatim in https://github.com/WordPress/gutenberg/pull/45472/commits/fc26baf7fd0f852122728c92db0b17c98c10f0c8

ciampo commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,116 @@
+	actions: Action[];
+	icon?: WPElement | null;
+	onDismiss?: Function;
+	status: string;
+	content: string;
+	__unstableHTML?: string;
+	isDismissible: boolean;
+	type: string;
```
```suggestion
	content: string;
	isDismissible: boolean;
```

Let's remove those `Notice` props, since they're currently not used in this component. We will add them back and type them properly when working on the `Notice` component

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/45472/commits/61f5caa469c97cc78cdccf4190be7b1472312c33

ciampo reviewed (commented):
THank you @kienstra for your work here!

I've left a few comments, I'll take another look at the code (including Storybook) once they've been addressed

Let's leave the work on the README last, so that we can sync it with whatever updates have been made to the component.

kienstra commented:
Thanks a lot for reviewing this, Marco! I'll apply your suggestions today.

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/snackbar/index.tsx

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/snackbar/types.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Paste Marco's TODO about moving WPNotice to the Notice component

kienstra committed:
Change spokenMessage and children to string

Thanks to Marco's
suggestion, as spokenMessage comes from
getNotices(), not React.

kienstra committed:
Change the type of icon to ReactNode

kienstra committed:
Apply Marco's suggestion for listRef's type

kienstra committed:
Commit Marco's removal of noop

kienstra committed:
Make the Action type be one of 2 types

kienstra committed:
Rename the type Action to NoticeAction

kienstra committed:
Replace the Function type with () => void

kienstra committed:
Replace the type ButtonEvent with SyntheticEvent

kienstra committed:
Add a story for SnackbarList

kienstra committed:
Replace SNACKBAR_REDUCE_MOTION_VARIANTS with undefined

kienstra committed:
Remove the single SnackbarList story

kienstra committed:
Change the SnackbarList children type to string

kienstra committed:
Remove DefaultTemplate, as it's only used once

kienstra committed:
Remove needless const savePostNotice

kienstra commented:
Hi @ciampo,
Thanks, your ideas made this a lot cleaner. 

Just finished all of your suggestions.

ciampo commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,101 @@
+	/**
+	 * Children to be rendered inside the notice list.
+	 */
+	children?: string;
```
The list's `children` prop should be a `ReactNode`

kienstra replied:
Sure, committed in https://github.com/WordPress/gutenberg/pull/45472/commits/a55a083aad978453a97d258cd9cf0a7f104102d8

ciampo commented on packages/components/src/snackbar/stories/index.tsx:
```diff
@@ -0,0 +1,96 @@
+	explicitDismiss: true,
+};
+
+export const WithActionAndExpicitDismiss: ComponentStory< typeof Snackbar > =
+	DefaultTemplate.bind( {} );
+WithActionAndExpicitDismiss.args = {
```
Typo

```suggestion
export const WithActionAndExplicitDismiss: ComponentStory< typeof Snackbar > =
	DefaultTemplate.bind( {} );
WithActionAndExplicitDismiss.args = {
```

ciampo commented on packages/components/src/snackbar/stories/index.tsx:
```diff
@@ -0,0 +1,96 @@
+);
+WithIcon.args = {
+	children: 'Add an icon to make your snackbar stand out',
+	icon: (
+		<span role="img" aria-label="Icon" style={ { fontSize: 21 } }>
+			🌮
+		</span>
+	),
```
It would be cool if we could use `@wordpress/icons` like we do in other components (eg. [`Button`](https://github.com/WordPress/gutenberg/blob/0ad21d5efcbdcc074806c6252d3836c85a6c0ee9/packages/components/src/button/stories/index.js)), but I gave it a quick try and it looks like those icons are not well supported by the component currently (layout and color especially).

One more thing to improve on this component after it gets refactored to TypeScript

kienstra replied:
Yeah, that'd be nice to simply have a `@wordpress/icon` as the icon.

ciampo commented on packages/components/src/snackbar/README.md:
````diff
@@ -36,17 +36,52 @@ const MySnackbarNotice = () => (
 ```
 
-#### Props
+### Props
````
Could we sort the props alphabetically?

kienstra replied:
Sure! Committed in https://github.com/WordPress/gutenberg/pull/45472/commits/84071c6354a66ce487477d632b64ee3e14cdf5c7

ciampo commented on packages/components/src/snackbar/README.md:
```diff
@@ -36,17 +36,52 @@ const MySnackbarNotice = () => (
+-   Required: No
+-   Default: `'polite'`
+
+#### `actions`: `Action[]`
+An array of action objects. Each member object should contain a `label` and either a `url` link string or `onClick` callback function.
+-   Required: No
+-   Default: `[]`
```
Some nitty formatting:

- let's add an empty line in between the prop's heading and the description paragraph
- let's add an empty line after the description paragraph and before the required/default part

```suggestion
#### `actions`: `Action[]`

An array of action objects. Each member object should contain a `label` and either a `url` link string or `onClick` callback function.

-   Required: No
-   Default: `[]`
```

Let's apply this formatting for all props

kienstra replied:
Ah, good idea. Committed in https://github.com/WordPress/gutenberg/pull/45472/commits/44a53712146af88dc39fe17e895b0fbb8d7d7cd7.

ciampo commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,101 @@
+	 * @default children
+	 */
+	spokenMessage?: Notice[ 'spokenMessage' ];
+	/**
+	 * A politeness level for the notice's spoken message. Should be provided as one of the valid options for [an `aria-live` attribute value](https://www.w3.org/TR/wai-aria-1.1/#aria-live). Defaults to `"polite"`. Note that this value should be considered a suggestion; assistive technologies may override it based on internal heuristics.
+	 *
+	 * A value of `'assertive'` is to be used for important, and usually time-sensitive, information. It will interrupt anything else the screen reader is announcing in that moment.
+	 * A value of `'polite'` is to be used for advisory information. It should not interrupt what the screen reader is announcing in that moment (the "speech queue") or interrupt the current task.
+	 *
+	 * @default 'polite'
+	 */
```
- no need to mention the default value in the description text, we already have a `default` tag
- let's not use markdown syntax for JSDoc links, instead we can use the `@see` notation
- let's keep each line within the 80 chars limit for improved legibility

```suggestion
	/**
	 * A politeness level for the notice's spoken message. Should be provided as
	 * one of the valid options for an `aria-live` attribute value. Note that this
	 * value should be considered a suggestion; assistive technologies may
	 * override it based on internal heuristics.
	 *
	 * A value of `'assertive'` is to be used for important, and usually
	 * time-sensitive, information. It will interrupt anything else the screen
	 * reader is announcing in that moment.
	 * A value of `'polite'` is to be used for advisory information. It should
	 * not interrupt what the screen reader is announcing in that moment
	 * (the "speech queue") or interrupt the current task.
	 *
	 * @see https://www.w3.org/TR/wai-aria-1.1/#aria-live
	 *
	 * @default 'polite'
	 */
```

kienstra replied:
Good ideas, these lines were long.

Committed in https://github.com/WordPress/gutenberg/pull/45472/commits/3939a0c8d6aec36f00786ca8c6e8e695846031c4 and https://github.com/WordPress/gutenberg/pull/45472/commits/70bfe595c8de58542f45de8ac4bb7b90fd31e75a

ciampo commented on packages/components/src/snackbar/types.ts:
```diff
@@ -0,0 +1,101 @@
+
+export type SnackbarProps = {
+	/**
+	 * The displayed message of a notice. Also used as the spoken message for assistive technology, unless `spokenMessage` is provided as an alternative message.
```
Let's break the line and make sure that comments in this file don't go beyond the 80 characters mark in each line (`children`, `actions` and `politeness`)

kienstra replied:
Sounds good, committed in https://github.com/WordPress/gutenberg/pull/45472/commits/70bfe595c8de58542f45de8ac4bb7b90fd31e75a

ciampo commented on packages/components/src/snackbar/README.md:
```diff
@@ -36,17 +36,52 @@ const MySnackbarNotice = () => (
+-   Default: `children`
+
+#### `politeness`: `'polite' | 'assertive'`
+A politeness level for the notice's spoken message. Should be provided as one of the valid options for [an `aria-live` attribute value](https://www.w3.org/TR/wai-aria-1.1/#aria-live). Defaults to `"polite"`. Note that this value should be considered a suggestion; assistive technologies may override it based on internal heuristics.
+-   A value of `'assertive'` is to be used for important, and usually time-sensitive, information. It will interrupt anything else the screen reader is announcing in that moment.
+-   A value of `'polite'` is to be used for advisory information. It should not interrupt what the screen reader is announcing in that moment (the "speech queue") or interrupt the current task.
```
- no need to mention the default value (it's already listed below)
- let's not use bullet points for the prop's description, as it will be joined with the required/default list below

```suggestion
A politeness level for the notice's spoken message. Should be provided as one of the valid options for [an `aria-live` attribute value](https://www.w3.org/TR/wai-aria-1.1/#aria-live). Note that this value should be considered a suggestion; assistive technologies may override it based on internal heuristics.

A value of `'assertive'` is to be used for important, and usually time-sensitive, information. It will interrupt anything else the screen reader is announcing in that moment.

A value of `'polite'` is to be used for advisory information. It should not interrupt what the screen reader is announcing in that moment (the "speech queue") or interrupt the current task.
```

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/45472/commits/4f823d39433212a35e9e37498745ae344b5328c4

ciampo reviewed (commented):
Thank you! I've left a few more comments.

---

More in general, there's a few issues that I've been noticing with the components, that are not related to this PR — we should aim at tackling those issues separately, in follow-up work:

 - `SnackbarList` is not very well-documented (it doesn't even have a README) and prior to this PR, didn't have Storybook examples either. We should re-organize the folder structure, improve the docs and come up with a better Storybook strategy (e.g. use subcomponents and merge `Snackbar` and `SnackbarList` in the same file)
 - We should  refactor the `Notice` component to TypeScript and re-use its types in `Snackbar`
 - I came across a framer-motion bug when clicking rapidly on two `Snackbar`s in a list — the second `Snackbar` goes transparent but doesn't get fully removed
 - The timeout logic for `Snackbar` to be dismissed works in a bit of an unexpected way (snackbar items don't get dismissed in order
 - Take `isDismissible` prop from notice data is not taken into account by `Snackbar` (although this could be on purpose)
 - The `Snackbar` icon (passed through its icon prop) is not vertically centered, doesn't seem to work well with WordPress icons

I went ahead and added those to the PR description, under a new "Follow up" section for now.

ciampo commented on packages/components/CHANGELOG.md:
```diff
@@ -56,6 +56,7 @@
 -   `Autocomplete`: use Popover's new `placement` prop instead of legacy `position` prop ([#44396](https://github.com/WordPress/gutenberg/pull/44396/)).
 -   `SelectControl`: Add `onChange`, `onBlur` and `onFocus` to storybook actions ([#45432](https://github.com/WordPress/gutenberg/pull/45432/)).
 -   `FontSizePicker`: Add more comprehensive tests ([#45298](https://github.com/WordPress/gutenberg/pull/45298)).
+-   `Snackbar`: Convert to TypeScript ([#45472](https://github.com/WordPress/gutenberg/pull/45472)).
```
Also, we should rebase this PR and move this CHANGELOG entry under "Unreleased", in the "### Internal" section

kienstra replied:
Ah, good idea. I just did that.

Is it alright that I didn't do `git rebase`? I just merged in `trunk` and resolved the conflict.

ciampo replied:
Sure, you can either merge or rebase — your preference.

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/snackbar/stories/index.tsx

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Move TODO to the Snackbar component

kienstra committed:
Commit Marco's edits to the SnackbarList storybook

kienstra committed:
Change the type of children in SnackbarList

kienstra committed:
Add empty lines in README.md

kienstra committed:
Alphabetize props

kienstra committed:
Remove markdown syntax, shorten JSDoc lines

kienstra committed:
Commit Marco's suggestion for README.md

kienstra committed:
Merge in trunk, resolve conflict

Move Snackbar to the bottom of CHANGELOG > Internal

kienstra committed:
Delete syntax from merge I forgot to delete

kienstra committed:
Restore CHANGELOG entries I deleted

kienstra committed:
Commit Marco's suggestion for the politeness prop

kienstra committed:
Remove and empty line between 2 lines

kienstra committed:
Remove reference to default, as there is one already

kienstra commented:
Hi @ciampo,
Thanks for your ideas and code. Just finished with your suggestions.

kienstra commented:
Also, I'll be out until Tuesday, November 8th.

kienstra committed:
Merge in trunk, resolve conflict in components/CHANGELOG.md

kienstra commented:
Hi @ciampo,
Would you have time for another round of review?

Thanks for your huge amount of patience on my PRs 😄

ciampo commented:
Hey @kienstra , I've been partially AFK last week and will be a bit slow to respond this week too — but I'll definitely get back to you!

ciampo commented on packages/components/src/snackbar/README.md:
```diff
@@ -36,17 +36,72 @@ const MySnackbarNotice = () => (
-    -   A value of `'polite'` is to be used for advisory information. It should not interrupt what the screen reader is announcing in that moment (the "speech queue") or interrupt the current task.
--   `onRemove`: function called when dismissing the notice.
--   `actions`: (array) an array of action objects. Each member object should contain a `label` and either a `url` link string or `onClick` callback function.
+#### `actions`: `Action[]`
```
```suggestion
#### `actions`: `NoticeAction[]`
```

kienstra replied:
Thanks, committed in https://github.com/WordPress/gutenberg/pull/45472/commits/8c33d6d519f45c40620692312e34020b39ab24ba

ciampo commented on packages/components/src/snackbar/README.md:
```diff
@@ -36,17 +36,72 @@ const MySnackbarNotice = () => (
+-   Required: No
+-   Default: `null`
+
+#### `listRef`: `MutableRefObject< HTMLDivElement >`
```
```suggestion
#### `listRef`: `MutableRefObject< HTMLDivElement | null >`
```

kienstra replied:
Committed in https://github.com/WordPress/gutenberg/pull/45472/commits/fe394e6149e7f63b4259ab5c5fbd145e78ef6b9e

ciampo commented on packages/components/src/snackbar/index.tsx:
```diff
@@ -14,21 +15,20 @@ import warning from '@wordpress/warning';
- *
- * @param {string|WPElement}     [message]  Message to announce.
- * @param {'polite'|'assertive'} politeness Politeness to announce.
  */
```
It's probably better to leave these JSDocs, we can just remove the types (since they're now defined in TypeScript)

```js
/**
 * Custom hook which announces the message with the given politeness, if a
 * valid message is provided.
 *
 * @param  message    Message to announce.
 * @param  politeness Politeness to announce.
 */
```

kienstra replied:
Sounds good, committed that verbatim: https://github.com/WordPress/gutenberg/pull/45472/commits/b145a587b113ed2381fbb606584e287fe9624794

ciampo reviewed (commented):
Just a few tweaks, and then we're good to merge! Thank you 🙏

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/snackbar/README.md

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Commit Marco's suggestion: Update packages/components/src/snackbar/README.md

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

kienstra committed:
Restore the JSDoc, using Marco's suggestion

kienstra committed:
Merge in trunk, resolve conflict

kienstra commented:
Thanks, @ciampo! Just applied your suggestions. 

No hurry responding, thanks for responding when you're AFK.

ciampo reviewed (approved):
LGTM 🚀 

Thank you for your work on this one!

kienstra commented:
Thanks so much, Marco! You're so patient.
